File: probe/crates/probe-remote/src/report.rs
//! Assemble an [`AssetReport`] from per-asset JSON pulled back from a remote scan.

use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{bail, Context};
use chrono::Utc;
use probe_contract::{Asset, AssetReport, HostInfo, Vulnerability};
use uuid::Uuid;

const MALWARE_JSON: &str = "malware.json";
const ASSET_JSON_FILES: &[&str] = &[
    "packages.json",
    "services.json",
    "accounts.json",
    "credentials.json",
];

/// Build an [`AssetReport`] from `host.json` / `packages.json` under `output_dir`.
///
/// `host.json` is required (use `--target host` or `all`). `packages.json` is
/// optional and merged into `assets` when present.
pub fn assemble_asset_report(output_dir: &Path) -> anyhow::Result<AssetReport> {
    let host_path = output_dir.join("host.json");
    if !host_path.is_file() {
        bail!(
            "host.json missing under {}; ingest requires --target host or all",
            output_dir.display()
        );
    }

    let host = read_json::<HostInfo>(&host_path)
        .with_context(|| format!("parse host.json at {}", host_path.display()))?;

    let assets = read_merged_assets(output_dir)?;

    Ok(AssetReport {
        report_id: format!("report-{}", Uuid::new_v4()),
        collected_at: Utc::now(),
        scanner_version: env!("CARGO_PKG_VERSION").to_string(),
        host,
        assets,
        vulnerabilities: Vec::new(),
    })
}

/// Like [`assemble_asset_report`], but merges `malware.json` when present.
pub fn finalize_asset_report(output_dir: &Path) -> anyhow::Result<AssetReport> {
    let mut report = assemble_asset_report(output_dir)?;
    attach_malware(&mut report, output_dir);
    Ok(report)
}

/// Merge ClamAV hits from `malware.json`, rebinding `affected_asset_id` to the host.
pub fn attach_malware(report: &mut AssetReport, output_dir: &Path) {
    let path = output_dir.join(MALWARE_JSON);
    let Ok(text) = fs::read_to_string(&path) else {
        return;
    };
    let Ok(mut vulns) = serde_json::from_str::<Vec<Vulnerability>>(&text) else {
        return;
    };
    let host_id = report.host.host_id.clone();
    for v in &mut vulns {
        v.affected_asset_id = host_id.clone();
    }
    report.vulnerabilities = vulns;
}

/// Write `asset_report.json` next to the pulled per-asset files.
pub fn write_asset_report(output_dir: &Path, report: &AssetReport) -> anyhow::Result<PathBuf> {
    let path = output_dir.join("asset_report.json");
    let file = fs::File::create(&path).with_context(|| format!("create {}", path.display()))?;
    serde_json::to_writer_pretty(file, report)
        .with_context(|| format!("write {}", path.display()))?;
    Ok(path)
}

fn read_merged_assets(output_dir: &Path) -> anyhow::Result<Vec<Asset>> {
    let mut assets = Vec::new();
    for fname in ASSET_JSON_FILES {
        let path = output_dir.join(fname);
        if !path.is_file() {
            continue;
        }
        let batch = read_json::<Vec<Asset>>(&path)
            .with_context(|| format!("parse {fname} at {}", path.display()))?;
        assets.extend(batch);
    }
    Ok(assets)
}

fn read_json<T: serde::de::DeserializeOwned>(path: &Path) -> anyhow::Result<T> {
    let text = fs::read_to_string(path).with_context(|| format!("read {}", path.display()))?;
    serde_json::from_str(&text).with_context(|| format!("decode JSON from {}", path.display()))
}

#[cfg(test)]
mod tests {
    use super::*;
    use probe_contract::Asset;

    fn sample_host() -> HostInfo {
        HostInfo {
            host_id: "host-demo-root".to_string(),
            hostname: "demo".to_string(),
            os: "Ubuntu 22.04".to_string(),
            kernel: None,
            arch: Some("x86_64".to_string()),
            ip_addrs: Vec::new(),
            mac_addrs: Vec::new(),
            boot_time: None,
        }
    }

    #[test]
    fn assembles_host_and_packages() {
        let dir = tempfile::tempdir().unwrap();
        fs::write(
            dir.path().join("host.json"),
            serde_json::to_string(&sample_host()).unwrap(),
        )
        .unwrap();
        fs::write(
            dir.path().join("packages.json"),
            serde_json::json!([{
                "kind": "package",
                "asset_id": "pkg-openssl",
                "name": "openssl",
                "version": "3.0.2",
                "source": "dpkg",
                "install_path": null,
                "ecosystem": "Ubuntu:22.04"
            }])
            .to_string(),
        )
        .unwrap();

        let report = assemble_asset_report(dir.path()).unwrap();
        assert_eq!(report.host.hostname, "demo");
        assert_eq!(report.assets.len(), 1);
        assert!(matches!(&report.assets[0], Asset::Package(p) if p.name == "openssl"));
        assert!(report.vulnerabilities.is_empty());
    }

    #[test]
    fn host_only_report_has_empty_assets() {
        let dir = tempfile::tempdir().unwrap();
        fs::write(
            dir.path().join("host.json"),
            serde_json::to_string(&sample_host()).unwrap(),
        )
        .unwrap();

        let report = assemble_asset_report(dir.path()).unwrap();
        assert!(report.assets.is_empty());
    }

    #[test]
    fn missing_host_json_is_error() {
        let dir = tempfile::tempdir().unwrap();
        fs::write(dir.path().join("packages.json"), "[]").unwrap();
        assert!(assemble_asset_report(dir.path()).is_err());
    }

    #[test]
    fn merges_malware_json_into_vulnerabilities() {
        let dir = tempfile::tempdir().unwrap();
        fs::write(
            dir.path().join("host.json"),
            serde_json::to_string(&sample_host()).unwrap(),
        )
        .unwrap();
        fs::write(
            dir.path().join("malware.json"),
            serde_json::json!([{
                "vuln_id": "Eicar-Test-Signature",
                "severity": "critical",
                "cvss_score": null,
                "affected_asset_id": "/tmp/eicar",
                "source": "clamav",
                "evidence": "infected file: /tmp/eicar",
                "references": [],
            }])
            .to_string(),
        )
        .unwrap();

        let report = finalize_asset_report(dir.path()).unwrap();
        assert_eq!(report.vulnerabilities.len(), 1);
        assert_eq!(
            report.vulnerabilities[0].affected_asset_id,
            "host-demo-root"
        );
        assert_eq!(report.vulnerabilities[0].vuln_id, "Eicar-Test-Signature");
    }

    #[test]
    fn writes_asset_report_json() {
        let dir = tempfile::tempdir().unwrap();
        fs::write(
            dir.path().join("host.json"),
            serde_json::to_string(&sample_host()).unwrap(),
        )
        .unwrap();

        let report = assemble_asset_report(dir.path()).unwrap();
        let path = write_asset_report(dir.path(), &report).unwrap();
        assert_eq!(path.file_name().unwrap(), "asset_report.json");
        let roundtrip: AssetReport =
            serde_json::from_str(&fs::read_to_string(&path).unwrap()).unwrap();
        assert_eq!(roundtrip.host.host_id, "host-demo-root");
    }
}

File: analyzer/src/analyzer/deploy/report.py
"""Assemble an :class:`AssetReport` from per-asset JSON pulled off a target.

The JSON files written by ``agent-host -o DIR`` mirror analyzer's own Pydantic
contracts, so we validate them directly with :mod:`analyzer.schemas` rather than a
separate Rust-side mirror.
"""

from __future__ import annotations

import os
import urllib.error
import urllib.request
import uuid
from datetime import UTC, datetime
from pathlib import Path

from pydantic import TypeAdapter

from ..schemas import Asset, AssetReport, HostInfo, Vulnerability

_ASSET_FILES = ("packages.json", "services.json", "accounts.json", "credentials.json")
_MALWARE_JSON = "malware.json"
_ASSET_LIST = TypeAdapter(list[Asset])
_VULN_LIST = TypeAdapter(list[Vulnerability])


def assemble_asset_report(output_dir: Path) -> AssetReport:
    """Build an AssetReport from ``host.json`` + per-asset files under ``output_dir``.

    ``host.json`` is required (use ``-t host`` or ``-t all``); the per-asset
    files are optional and merged into ``assets`` when present.
    """
    host_path = output_dir / "host.json"
    if not host_path.is_file():
        raise FileNotFoundError(
            f"host.json missing under {output_dir}; assembly requires -t host or all"
        )
    host = HostInfo.model_validate_json(host_path.read_text(encoding="utf-8"))

    assets: list[Asset] = []
    for fname in _ASSET_FILES:
        path = output_dir / fname
        if path.is_file():
            assets.extend(_ASSET_LIST.validate_json(path.read_text(encoding="utf-8")))

    return AssetReport(
        report_id=f"report-{uuid.uuid4()}",
        collected_at=datetime.now(UTC),
        scanner_version="analyzer-scan/0.1",
        host=host,
        assets=assets,
        vulnerabilities=[],
    )


def attach_malware(report: AssetReport, output_dir: Path) -> None:
    """Merge malware hits from ``malware.json``, rebinding ``affected_asset_id`` to the host."""
    path = output_dir / _MALWARE_JSON
    if not path.is_file():
        return
    try:
        vulns = _VULN_LIST.validate_json(path.read_text(encoding="utf-8"))
    except Exception:  # noqa: BLE001 - a malformed malware.json must not break assembly
        return
    for vuln in vulns:
        vuln.affected_asset_id = report.host.host_id
    report.vulnerabilities = vulns


def finalize_asset_report(output_dir: Path) -> AssetReport:
    """:func:`assemble_asset_report` plus :func:`attach_malware` when present."""
    report = assemble_asset_report(output_dir)
    attach_malware(report, output_dir)
    return report


def write_asset_report(output_dir: Path, report: AssetReport) -> Path:
    """Write ``asset_report.json`` next to the pulled per-asset files."""
    path = output_dir / "asset_report.json"
    path.write_text(report.model_dump_json(indent=2), encoding="utf-8")
    return path


def upload_asset_report(report: AssetReport, form_base_url: str) -> None:
    """POST the report to analyzer's ``/ingest/asset-report`` (ANALYZER_API_TOKEN bearer)."""
    url = form_base_url.strip().rstrip("/") + "/ingest/asset-report"
    body = report.model_dump_json().encode("utf-8")
    request = urllib.request.Request(url, data=body, method="POST")
    request.add_header("Content-Type", "application/json")
    token = os.environ.get("ANALYZER_API_TOKEN")
    if token:
        request.add_header("Authorization", f"Bearer {token}")
    try:
        with urllib.request.urlopen(request, timeout=60) as response:
            status = response.status
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", "replace")
        raise RuntimeError(f"analyzer ingest failed ({exc.code}): {detail}") from exc
    if status != 202:
        raise RuntimeError(f"analyzer ingest returned unexpected status {status}")


def upload_flow_batch(flow_json: Path, analyzer_base_url: str) -> None:
    """POST a pulled `FlowBatch` JSON file to analyzer's ``/ingest/flow-batch``."""
    url = analyzer_base_url.strip().rstrip("/") + "/ingest/flow-batch"
    body = flow_json.read_bytes()
    request = urllib.request.Request(url, data=body, method="POST")
    request.add_header("Content-Type", "application/json")
    token = os.environ.get("ANALYZER_API_TOKEN")
    if token:
        request.add_header("Authorization", f"Bearer {token}")
    try:
        with urllib.request.urlopen(request, timeout=60) as response:
            status = response.status
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", "replace")
        raise RuntimeError(f"analyzer flow ingest failed ({exc.code}): {detail}") from exc
    if status != 202:
        raise RuntimeError(f"analyzer flow ingest returned unexpected status {status}")
